File: 2022/06/08/kys6453/yongsu.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans = 0
        tmp_ans = 0
        last_duplicate = 0
        word_dict = {} # 문자열이 나왔는지 확인하기 위한 hashmap
        cnt = 0
        for c in s: # 문자열 순회
            if c in word_dict: # 만약 현재 문자가 이미 나왔던 문자라면
                if last_duplicate > word_dict[c][1]: # 만약 마지막으로 중복되었다고 판단한 문자가 현재 문자가 나왔던 적보다 나중이라면 거기부터 다시 시작해야함
                    tmp_ans = cnt - last_duplicate # 현재 위치부터 마지막으로 중복되었다고 판단된 문자까지의 길이를 구함
                else: # 만약 현재 문자가 나왔던 적이 마지막으로 중복되었다고 저장했던 문자보다 이후라면 현재 문자가 나왔던 곳부터 다시 시작해야함
                    tmp_ans = cnt - word_dict[c][1]  # 현재 위치부터 해당 문자가 나왔던 적까지의 길이를 구함
                    last_duplicate = word_dict[c][1] # 그리고 마지막으로 중복되었다고 판단된 문자의 인덱스 갱신
            else: # 만약 새로운 문자라면
                tmp_ans += 1 # 계속 문자열의 길이를 셈
            word_dict[c] = [1, cnt] # 문자열에 1이라고 체크한 후 해당 문자열의 인덱스를 함께 기록
            cnt += 1 # 인덱스를 늘린다
            ans = max(ans, tmp_ans) # 내가 마지막으로 저장한 답 후보와 새로 생긴 답 후보 중 더 긴 값을 return 할거임
        
        return ans


#Runtime: 129 ms, faster than 29.55% of Python3 online submissions for Longest Substring Without Repeating Characters.
#Memory Usage: 14.1 MB, less than 13.70% of Python3 online submissions for Longest Substring Without Repeating Characters.

File: 2022/06/13_zigzag_conversion/yongsu.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

char board[2000][2000];
int row_column_index[2000];
char ans[2000];

void init()
{
    int i, j;
    for (i = 0; i < 2000; i++)
    {
        memset(board[i], 0, strlen(board[i]));
        row_column_index[i] = 0;
    }
    for (i = 0; i < 2000; i++)
    {
        ans[i] = 0;
    }
}

char * convert(char * s, int numRows)
{
    int i, j;
    int len = strlen(s);

    int row_increment = -1; // 1 for ++, -1 for --
    int row_num = 0;

    init();

    if (numRows == 1) return s;

    for (i = 0; i < len; i++)
    {
        board[row_num][row_column_index[row_num]] = s[i];
        row_column_index[row_num]++;
        if (row_num == numRows - 1 || row_num == 0) row_increment = -row_increment;
        if (row_increment == 1) row_num++;
        else row_num--;
    }

    int cnt = 0;

    for (i = 0; i < numRows; i++)
    {
        for (j = 0; j < len; j++)
        {
            if (board[i][j] == 0) break;
            ans[cnt++] = board[i][j];
        }
    }

    return ans;
}

// int main(void)
// {
//     printf("hello world\n");
//     printf("%s\n", convert("PAYPALISHIRING", 3));
//     printf("%s\n", convert("PAYPALISHIRING", 4));
//     printf("%s\n", convert("AB", 1));
//     printf("%s\n", convert("uhyxutlx", 8));
//     return 0;
// }


// Runtime: 50 ms, faster than 15.20% of C online submissions for Zigzag Conversion.
// Memory Usage: 8.4 MB, less than 16.63% of C online submissions for Zigzag Conversion.